File: alerts/telegram_bot.py
import os
import logging
import asyncio
from telegram import Bot, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.error import TelegramError
from dotenv import load_dotenv
from agents.position_monitor import Alert

load_dotenv()

# Setup logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler("logs/telegram_bot.log"),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger("TelegramBot")

def format_alert(alert: Alert) -> str:
    emoji = {"INFO": "🟢", "WARNING": "🟡", "CRITICAL": "🔴"}.get(alert.level, "⚪")
    header = f"{emoji} {alert.level} — {alert.symbol}" if alert.symbol else f"{emoji} {alert.level}"
    msg = [
        header,
        "━━━━━━━━━━━━━━━━━",
        alert.message
    ]
    if alert.action:
        msg.append(f"Action: {alert.action}")
    msg.append("━━━━━━━━━━━━━━━━━")
    return "\n".join(msg)

async def send_telegram_alert(alert: Alert, retries: int = 3):
    token = os.getenv("TELEGRAM_BOT_TOKEN")
    chat_id = os.getenv("TELEGRAM_CHAT_ID")
    
    if not token or not chat_id:
        logger.warning(f"Telegram not configured. Alert: {alert.message}")
        return

    bot = Bot(token=token)
    text = format_alert(alert)
    
    reply_markup = None
    if alert.level == "CRITICAL":
        keyboard = [
            [
                InlineKeyboardButton("Acknowledge", callback_data=f"ack_{alert.symbol}"),
                InlineKeyboardButton("Run Review", callback_data=f"review_{alert.symbol}")
            ]
        ]
        reply_markup = InlineKeyboardMarkup(keyboard)

    for attempt in range(retries):
        try:
            await bot.send_message(
                chat_id=chat_id,
                text=text,
                reply_markup=reply_markup
            )
            logger.info(f"Alert sent to Telegram: {alert.level} - {alert.symbol}")
            return
        except TelegramError as e:
            logger.error(f"Attempt {attempt + 1} failed to send Telegram alert: {e}")
            if attempt < retries - 1:
                await asyncio.sleep(2 ** attempt) # Exponential backoff
            else:
                # Fallback log
                logger.critical(f"FALLBACK: Failed to send alert after {retries} attempts. Alert content: {text}")

File: src/alerts/telegramBot.ts
import { Telegraf } from 'telegraf';
import { Alert } from '../types/index.js';
import { logAlert } from '../db/database.js';
import { logger } from '../utils/logger.js';
import dotenv from 'dotenv';

dotenv.config();

const token = process.env.TELEGRAM_BOT_TOKEN;
const chatId = process.env.TELEGRAM_CHAT_ID;

const bot = token ? new Telegraf(token) : null;

/**
 * Sends an alert message to a Telegram chat and logs it to the database.
 * @param alert The alert to send.
 */
export async function sendTelegramAlert(alert: Alert): Promise<void> {
  if (!bot || !chatId) {
    logger.warn({ alert }, 'Telegram bot not configured');
    // Even if bot is not configured, we might want to log it to DB
    logAlert(alert);
    return;
  }

  const emoji = {
    'INFO': '🟢',
    'WARNING': '🟡',
    'CRITICAL': '🔴'
  }[alert.level];

  const message = `${emoji} *${alert.level}* ${alert.symbol ? `— ${alert.symbol}` : ''}\n${alert.message}${alert.action ? `\n\n*Action:* ${alert.action}` : ''}`;

  const extra: { 
    parse_mode: 'Markdown', 
    reply_markup?: { inline_keyboard: { text: string; callback_data: string }[][] } 
  } = { parse_mode: 'Markdown' };

  // Spec 6.9: Add basic inline keyboard support for CRITICAL alerts
  if (alert.level === 'CRITICAL') {
    extra.reply_markup = {
      inline_keyboard: [
        [
          { text: '✅ Acknowledge', callback_data: 'ack_alert' },
          { text: '🔄 Run Rebalancing', callback_data: 'run_rebalancing' }
        ]
      ]
    };
  }

  try {
    await bot.telegram.sendMessage(chatId, message, extra);
    // Log alert to DB
    logAlert(alert);
  } catch (error) {
    logger.error(error, 'Failed to send Telegram message');
    // Still try to log to DB if sending failed
    logAlert(alert);
  }
}
